automates/model_assembly/networks.py

```diff
@@ -819,10 +855,25 @@ def __init__(
         self.scope = id.scope
         self.name = id.con_name
         self.label = f"{self.name} ({self.namespace}.{self.scope})"
+        self.metadata = M
 
         self.variables = [n for n in self.nodes if isinstance(n, VariableNode)]
         self.lambdas = [n for n in self.nodes if isinstance(n, LambdaNode)]
-        self.types = []
+        self.types = T
+
+        # NOTE: removing detached variables from GrFN
+        del_indices = list()
+        for idx, var_node in enumerate(self.variables):
+            found_var = False
+            for edge in self.hyper_edges:
+                if var_node in edge.inputs or var_node in edge.outputs:
+                    found_var = True
+                    break
+            if not found_var:
+                self.remove_node(var_node)
+                del_indices.append(idx)
```

Comment:
why don't you just delete the variable from self.variables here? (I think its because you cant modify the list while looping through it, but does enumerate create make a copy of this list so it doesn't matter?)

Reply:
Enumerate is only creating an iterator for the (mutable) object that uses the objects included `__next__()` method not a copy of the object and as you point out you don't want to modify a Python iterable during iteration.